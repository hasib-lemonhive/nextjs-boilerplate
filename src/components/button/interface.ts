import React from 'react';
import { IconNames } from './../icons/interface';

export const buttonColorSchemes = [
  'primary-bg-text-white',
  'secondary-bg-text-white',
  'black-bg-text-white',
  'white-bg-text-black',
] as const;
type ColorScheme = (typeof buttonColorSchemes)[number];

export const outlineButtonColorSchemes = [
  'transparent-bg-text-primary',
  'transparent-bg-text-secondary',
  'transparent-bg-text-white',
  'transparent-bg-text-black',
] as const;
type OutlineColorScheme = (typeof outlineButtonColorSchemes)[number];

export const buttonSizes = ['lg', 'md', 'sm', 'xs'] as const;
type Size = (typeof buttonSizes)[number];

export interface IFillButton {
  label: string;
  colorScheme: ColorScheme;
  size: Size;
  clickHandler: (
    e: React.MouseEvent<HTMLElement> | React.FormEvent<HTMLFormElement>
  ) => void;
  hoverColorScheme?: ColorScheme;
  iconName?: IconNames;
  iconIsLeft?: boolean;
  shiftIconOnHover?: boolean;
  isDisabled?: boolean;
  variant: 'fill';
}

export interface IFillRedirectButton
  extends Omit<IFillButton, 'clickHandler' | 'isDisabled'> {
  href: string;
  isOpenNewTab: boolean;
}

export interface IOutlineButton
  extends Omit<IFillButton, 'colorScheme' | 'hoverColorScheme' | 'variant'> {
  colorScheme: OutlineColorScheme;
  hoverColorScheme: OutlineColorScheme;
  variant: 'outline';
}

export interface IOutlineRedirectButton
  extends Omit<IOutlineButton, 'clickHandler' | 'isDisabled'> {
  href: string;
  isOpenNewTab: boolean;
}

export type IBaseButton =
  | Omit<IOutlineButton, 'clickHandler'>
  | Omit<IFillButton, 'clickHandler'>;
