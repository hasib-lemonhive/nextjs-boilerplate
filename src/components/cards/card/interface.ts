import { ISanityCustomPortableText } from 'src/backend/types/general/portable-text';

export const CardImageDimensions = {
  height: 400,
  width: 600,
};

export interface ICard {
  item: {
    name: string;
    description: ISanityCustomPortableText;
    image: {
      src: string;
      alt: string;
      lqip: string;
    };
  };
  link?: string;
}
