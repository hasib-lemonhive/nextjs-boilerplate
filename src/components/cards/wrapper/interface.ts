export interface ICardWrapper {
  dataTestId: string;
  cardShadow: CardShadowEnum;
  hoverCardShadow?: CardShadowEnum;
  cardBorderRadius: CardBorderRadiusEnum;
}

export enum CardShadowEnum {
  ShadowOne = 'shadow-one',
  None = 'shadow-none',
}

export enum CardBorderRadiusEnum {
  RadiusOne = 'radius-one',
  RadiusTwo = 'radius-two',
  None = 'radius-none',
}
