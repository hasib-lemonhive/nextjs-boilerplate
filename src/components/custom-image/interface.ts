export interface ICustomImage {
  src: string;
  height: number;
  width: number;
  alt: string;
  lqip?: string;
}

export interface ICustomFillImage {
  src: string;
  alt: string;
  lqip?: string;
}
