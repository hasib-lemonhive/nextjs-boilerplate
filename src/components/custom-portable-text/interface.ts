import { ISanityCustomPortableText } from 'src/backend/types/general/portable-text';

export interface ICustomPortableText {
  content: ISanityCustomPortableText;
}
