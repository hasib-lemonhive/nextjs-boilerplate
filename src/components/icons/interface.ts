type ArrowIcons = 'Arrow Left' | 'Arrow Right';
type ChevronIcons = 'Chevron Left' | 'Chevron Right';
type VectorIcons = 'Error Vector' | 'No Data Vector';

export const icons = ['Train', 'Airplane', 'Rocket'] as const;
type Icons = (typeof icons)[number];

export type IconNames = Icons | ArrowIcons | ChevronIcons | VectorIcons;
