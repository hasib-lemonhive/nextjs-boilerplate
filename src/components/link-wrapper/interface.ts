import { ReactNode } from 'react';

export interface ILinkWrapper {
  href: string;
  children: ReactNode;
  isOpenNewTab: boolean;
  isStyled?: boolean;
}
