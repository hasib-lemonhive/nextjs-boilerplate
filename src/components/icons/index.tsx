import { IconNames } from './interface';
import { ErrorVector } from './sub-components/error';
import { NoDataVector } from './sub-components/no-data';
import { FaArrowLeft, FaArrowRight } from 'react-icons/fa';

interface IIconStoreProps {
  iconName: IconNames;
}

/**
 * @method IconStore
 * @param {iconName} IIconStoreProps
 * @returns {JSX.Element}
 * @description this function will receive iconName and return icon JSX.Element(svg)
 */
const IconStore = ({ iconName }: IIconStoreProps) => {
  switch (iconName) {
    case 'error-vector': {
      return <ErrorVector />;
    }
    case 'no-data-vector': {
      return <NoDataVector />;
    }
    case 'arrow-left': {
      return <FaArrowLeft />;
    }
    case 'arrow-right': {
      return <FaArrowRight />;
    }
    case 'chevron-left': {
      return <FaArrowLeft />;
    }
    case 'chevron-right': {
      return <FaArrowRight />;
    }
  }
};

export default IconStore;
