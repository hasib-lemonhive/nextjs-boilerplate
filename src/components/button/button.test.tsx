import { render, fireEvent } from '@testing-library/react';
import Button from '.';
import {
  buttonProps,
  outlineButtonProps,
  redirectButtonProps,
} from './mock-data';
import {
  buttonColorSchemes,
  buttonSizes,
  outlineButtonColorSchemes,
} from './interface';
import { icons } from '@components/icons/interface';

it('renders button correctly', () => {
  const { getByRole } = render(<Button {...buttonProps} />);
  const element = getByRole('button');
  expect(element).toBeInTheDocument();
  expect(element).toMatchSnapshot();
});

it('renders redirect button correctly', () => {
  const { getByRole } = render(<Button.Redirect {...redirectButtonProps} />);
  const element = getByRole('link');
  expect(element).toBeInTheDocument();
  expect(element).toMatchSnapshot();
});

describe('Given a mock function is passed to button', () => {
  it('Expect mockFunction to be called on button click', () => {
    const mockFunction = jest.fn();

    const props = { ...buttonProps, clickHandler: mockFunction };
    const { getByRole } = render(<Button {...props} />);
    const button = getByRole('button');
    fireEvent.click(button);
    expect(mockFunction).toHaveBeenCalled();
  });
});

describe.each([true, false])(
  'Given open new tab configurations',
  (isOpenNewTab) => {
    it(`Expect redirect button to contain href and ${
      isOpenNewTab ? 'target=_blank' : 'target=_self'
    }`, () => {
      const props = {
        ...buttonProps,
        href: 'https://www.google.com/',
        isOpenNewTab: isOpenNewTab,
      };
      const { getByRole } = render(<Button.Redirect {...props} />);
      const link = getByRole('link');

      expect(link).toHaveAttribute('href', 'https://www.google.com/');

      if (isOpenNewTab) {
        expect(link).toHaveAttribute('target', '_blank');
      } else {
        expect(link).toHaveAttribute('target', '_self');
      }
    });
  }
);

describe.each(buttonSizes)('Given button size: %s', (size) => {
  it(`Expect button to have class ${size}`, () => {
    const props = { ...buttonProps, size: size };
    const { getByRole } = render(<Button {...props} />);
    const button = getByRole('button');
    expect(button.children[0]).toHaveClass(size);
  });
});

describe.each(icons)('Given icon option: %s', (iconName) => {
  it(`Expect button to have ${iconName}`, () => {
    const props = { ...buttonProps, iconName: iconName };
    const { getByRole } = render(<Button {...props} />);
    const button = getByRole('button');
    expect(button).toMatchSnapshot();
  });
});

describe.each(buttonColorSchemes)('Given color scheme %s', (colorScheme) => {
  it(`Expect button to have class ${colorScheme}`, () => {
    const props = { ...buttonProps, colorScheme: colorScheme };
    const { getByRole } = render(<Button {...props} />);
    const button = getByRole('button');

    expect(button.children[0]).toHaveClass(colorScheme);
  });
});

describe.each(outlineButtonColorSchemes)(
  'Given color scheme %s',
  (colorScheme) => {
    it(`Expect outline button to have class ${colorScheme}`, () => {
      const props = { ...outlineButtonProps, colorScheme: colorScheme };
      const { getByRole } = render(<Button {...props} />);
      const button = getByRole('button');

      expect(button.children[0]).toHaveClass(colorScheme);
    });
  }
);

describe.each([true, false])('Given iconIsLeft boolean', (iconIsLeft) => {
  it(`Expect button to ${
    iconIsLeft ? 'not' : ''
  } contain class: flex-row-reverse`, () => {
    const props = { ...buttonProps, iconIsLeft: iconIsLeft };
    const { getByRole } = render(<Button {...props} iconName="Arrow Right" />);
    const button = getByRole('button');

    if (iconIsLeft) {
      expect(button.children[0]).toHaveClass('flex-row-reverse');
    } else {
      expect(button.children[0]).not.toHaveClass('flex-row-reverse');
    }
  });
});

describe.each([true, false])(
  'Given shiftIconOnHover boolean',
  (shiftIconOnHover) => {
    it(`Expect button to ${
      shiftIconOnHover ? 'not' : ''
    } contain class: shift-icon-on-hover`, () => {
      const props = { ...buttonProps, shiftIconOnHover: shiftIconOnHover };
      const { getByRole } = render(
        <Button {...props} iconName="Arrow Right" />
      );
      const button = getByRole('button');

      if (shiftIconOnHover) {
        expect(button.children[0]).toHaveClass('shift-icon-on-hover');
      } else {
        expect(button.children[0]).not.toHaveClass('shift-icon-on-hover');
      }
    });
  }
);
