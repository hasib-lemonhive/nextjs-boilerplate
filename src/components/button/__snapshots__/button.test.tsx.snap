// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given icon option: Airplane Expect button to have Airplane 1`] = `
<button
  class="inline-flex focus:outline-none focus-visible:outline-none focus:shadow-none focus-visible:shadow-none"
  type="button"
>
  <span
    aria-label="Button"
    class="relative z-10 group/button overflow-hidden inline-flex items-center justify-center gap-3 cursor-pointer text-white hover:text-white leading-[1.15] button primary-bg-text-white hover-secondary-bg-text-white lg shift-icon-on-hover"
  >
    <span>
      Button
    </span>
    <span
      class="button-icon"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M14 8.947L22 14v2l-8-2.526v5.36l3 1.666V22l-4.5-1L8 22v-1.5l3-1.667v-5.36L3 16v-2l8-5.053V3.5a1.5 1.5 0 0 1 3 0v5.447z"
          />
        </g>
      </svg>
    </span>
    <span
      class="absolute z-[-1] inset-0 invisible opacity-0 group-hover/button:visible group-hover/button:opacity-100 button-bg"
    />
  </span>
</button>
`;

exports[`Given icon option: Rocket Expect button to have Rocket 1`] = `
<button
  class="inline-flex focus:outline-none focus-visible:outline-none focus:shadow-none focus-visible:shadow-none"
  type="button"
>
  <span
    aria-label="Button"
    class="relative z-10 group/button overflow-hidden inline-flex items-center justify-center gap-3 cursor-pointer text-white hover:text-white leading-[1.15] button primary-bg-text-white hover-secondary-bg-text-white lg shift-icon-on-hover"
  >
    <span>
      Button
    </span>
    <span
      class="button-icon"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M461.8 53.6c-.4-1.7-1.6-3-3.3-3.4-54.4-13.3-180.1 34.1-248.2 102.2-13.3 13.3-24.2 26.4-33.1 39.1-21-1.9-42-.3-59.9 7.5-50.5 22.2-65.2 80.2-69.3 105.1-1 5.9 3.9 11 9.8 10.4l81.1-8.9c.1 7.8.6 14 1.1 18.3.4 4.2 2.3 8.1 5.3 11.1l31.4 31.4c3 3 6.9 4.9 11.1 5.3 4.3.5 10.5 1 18.2 1.1l-8.9 81c-.6 5.9 4.5 10.8 10.4 9.8 24.9-4 83-18.7 105.1-69.2 7.8-17.9 9.4-38.8 7.6-59.7 12.7-8.9 25.9-19.8 39.2-33.1 68.4-68 115.5-190.9 102.4-248zM298.6 213.5c-16.7-16.7-16.7-43.7 0-60.4 16.7-16.7 43.7-16.7 60.4 0 16.7 16.7 16.7 43.7 0 60.4-16.7 16.7-43.7 16.7-60.4 0z"
        />
        <path
          d="M174.5 380.5c-4.2 4.2-11.7 6.6-19.8 8-18.2 3.1-34.1-12.8-31-31 1.4-8.1 3.7-15.6 7.9-19.7l.1-.1c2.3-2.3.4-6.1-2.8-5.7-9.8 1.2-19.4 5.6-26.9 13.1-18 18-19.7 84.8-19.7 84.8s66.9-1.7 84.9-19.7c7.6-7.6 11.9-17.1 13.1-26.9.3-3.2-3.6-5.1-5.8-2.8z"
        />
      </svg>
    </span>
    <span
      class="absolute z-[-1] inset-0 invisible opacity-0 group-hover/button:visible group-hover/button:opacity-100 button-bg"
    />
  </span>
</button>
`;

exports[`Given icon option: Train Expect button to have Train 1`] = `
<button
  class="inline-flex focus:outline-none focus-visible:outline-none focus:shadow-none focus-visible:shadow-none"
  type="button"
>
  <span
    aria-label="Button"
    class="relative z-10 group/button overflow-hidden inline-flex items-center justify-center gap-3 cursor-pointer text-white hover:text-white leading-[1.15] button primary-bg-text-white hover-secondary-bg-text-white lg shift-icon-on-hover"
  >
    <span>
      Button
    </span>
    <span
      class="button-icon"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M17.2 20l1.8 1.5v.5H5v-.5L6.8 20H5a2 2 0 0 1-2-2V7a4 4 0 0 1 4-4h10a4 4 0 0 1 4 4v11a2 2 0 0 1-2 2h-1.8zM5 7v4h14V7H5zm7 11a2 2 0 1 0 0-4 2 2 0 0 0 0 4z"
          />
        </g>
      </svg>
    </span>
    <span
      class="absolute z-[-1] inset-0 invisible opacity-0 group-hover/button:visible group-hover/button:opacity-100 button-bg"
    />
  </span>
</button>
`;

exports[`renders button correctly 1`] = `
<button
  class="inline-flex focus:outline-none focus-visible:outline-none focus:shadow-none focus-visible:shadow-none"
  type="button"
>
  <span
    aria-label="Button"
    class="relative z-10 group/button overflow-hidden inline-flex items-center justify-center gap-3 cursor-pointer text-white hover:text-white leading-[1.15] button primary-bg-text-white hover-secondary-bg-text-white lg"
  >
    <span>
      Button
    </span>
    <span
      class="absolute z-[-1] inset-0 invisible opacity-0 group-hover/button:visible group-hover/button:opacity-100 button-bg"
    />
  </span>
</button>
`;

exports[`renders redirect button correctly 1`] = `
<a
  class="inline-flex focus:outline-none focus-visible:outline-none focus:shadow-none focus-visible:shadow-none"
  href="https://www.google.com/"
  target="_blank"
>
  <span
    aria-label="Redirect Button"
    class="relative z-10 group/button overflow-hidden inline-flex items-center justify-center gap-3 cursor-pointer text-white hover:text-white leading-[1.15] button primary-bg-text-white hover-secondary-bg-text-white lg"
  >
    <span>
      Redirect Button
    </span>
    <span
      class="absolute z-[-1] inset-0 invisible opacity-0 group-hover/button:visible group-hover/button:opacity-100 button-bg"
    />
  </span>
</a>
`;
